src/gui/qgsfilepickerwidget.cpp

```diff
@@ -0,0 +1,265 @@
+  connect( mLineEdit, SIGNAL( textEdited( QString ) ), this, SLOT( textEdited( QString ) ) );
+  layout->addWidget( mLineEdit, 1, 0 );
+
+  mFilePickerButton = new QToolButton( this );
+  mFilePickerButton->setText( "..." );
+  connect( mFilePickerButton, SIGNAL( clicked() ), this, SLOT( openFileDialog() ) );
+  layout->addWidget( mFilePickerButton, 0, 1, 2, 1 );
+
+  setLayout( layout );
+}
+
+QString QgsFilePickerWidget::filePath()
+{
+  return mFilePath;
+}
+
+void QgsFilePickerWidget::setFilePath( QString path )
+{
+  mFilePath = path;
+  mLineEdit->setText( path );
+  mLinkLabel->setText( toUrl( path ) );
+  emit fileChanged( mFilePath );
+}
+
+void QgsFilePickerWidget::setReadOnly( bool readOnly )
+{
+  mFilePickerButton->setEnabled( !readOnly );
+  mLineEdit->setEnabled( !readOnly );
+}
+
+bool QgsFilePickerWidget::filePickerButtonVisible() const
+{
+  return mButtonVisible;
+}
+
+void QgsFilePickerWidget::setFilePickerButtonVisible( bool visible )
+{
+  mButtonVisible = visible;
+  mFilePickerButton->setVisible( visible );
+}
+
+void QgsFilePickerWidget::textEdited( QString path )
+{
+  mFilePath = path;
+  mLinkLabel->setText( toUrl( path ) );
+  emit fileChanged( mFilePath );
+}
+
+bool QgsFilePickerWidget::useLink() const
+{
+  return mUseLink;
+}
+
+void QgsFilePickerWidget::setUseLink( bool useLink )
+{
+  mUseLink = useLink;
+  mLinkLabel->setVisible( mUseLink );
+  mLineEdit->setVisible( !mUseLink );
+}
+
+bool QgsFilePickerWidget::fullUrl() const
+{
+  return mFullUrl;
+}
+
+void QgsFilePickerWidget::setFullUrl( bool fullUrl )
+{
+  mFullUrl = fullUrl;
+}
+
+QString QgsFilePickerWidget::defaultRoot() const
+{
+  return mDefaultRoot;
+}
+
+void QgsFilePickerWidget::setDefaultRoot( QString defaultRoot )
+{
+  mDefaultRoot = defaultRoot;
+}
+
+QgsFilePickerWidget::StorageMode QgsFilePickerWidget::storageMode() const
+{
+  return mStorageMode;
+}
+
+void QgsFilePickerWidget::setStorageMode( QgsFilePickerWidget::StorageMode storageMode )
+{
+  mStorageMode = storageMode;
+}
+
+QgsFilePickerWidget::RelativeStorage QgsFilePickerWidget::relativeStorage()
+{
+  return mRelativeStorage;
+}
+
+void QgsFilePickerWidget::setRelativeStorage( QgsFilePickerWidget::RelativeStorage relativeStorage )
+{
+  mRelativeStorage = relativeStorage;
+}
+
+void QgsFilePickerWidget::openFileDialog()
+{
+  QSettings settings;
+  QString oldPath;
+
+  // If we use fixed default path
+  if ( !mDefaultRoot.isEmpty() )
+  {
+    oldPath = QDir::cleanPath( mDefaultRoot );
+  }
+  // if we use a relative path option, we need to obtain the full path
+  else if ( !mFilePath.isEmpty() )
+  {
+    oldPath = relativePath( mFilePath, false );
+  }
+
+  // If there is no valid value, find a default path to use
+  QUrl theUrl = QUrl::fromUserInput( oldPath );
+  if ( !theUrl.isValid() )
+  {
+    oldPath = settings.value( "/UI/lastExternalResourceWidgetDir", QDir::cleanPath( QgsProject::instance()->fileInfo().absolutePath() ) ).toString();
```

Comment:
What's the default path here if a project hasn't been saved? It should default to something sensible in this case, eg the users' documents folder (and definitely not the annoying default qgis bin folder!)